### PR TITLE
Adding CryptoJobMetrics and Tracker to track detailed crypto job metrics

### PR DESCRIPTION
**Pull request by nsivabalan**
This patch adds detailed cryptoJob metrics like request queuing time, processing time, response waiting time and so on. 

Coding style applied
SLA: 20 mins

### CODE REVIEW

File: ambry-router/src/main/java/com.github.ambry.router/DecryptJob.java [hunk 1]
```diff
@@ -62,6 +65,7 @@
    * 5. Invoke callback with the decryptedBlobContent
    */
   public void run() {
+    decryptJobMetricsTracker.startRequestProcessing();
```
Comment (vgkholla) [inline ambry-router/src/main/java/com.github.ambry.router/DecryptJob.java]: I'm not sure if "request" is the right term to use here.

File: ambry-router/src/main/java/com.github.ambry.router/GetBlobInfoOperation.java [hunk 2]
```diff
@@ -335,27 +337,32 @@ private void handleBody(InputStream payload, MessageMetadata messageMetadata)
...
-                  logger.trace("Exception {} thrown on decryption for {}", exception, blobId);
-                  setOperationException(
-                      new RouterException("Exception thrown on decrypting the content for " + blobId, exception,
-                          RouterErrorCode.UnexpectedInternalError));
-                  progressTracker.setDecryptionFailed();
-                }
-                routerCallback.onPollReady();
-              }));
+              decryptJobMetricsTracker, (DecryptJob.DecryptJobResult result, Exception exception) -> {
+            decryptJobMetricsTracker.startResponseProcessing();
```
Comment (vgkholla) [inline ambry-router/src/main/java/com.github.ambry.router/GetBlobInfoOperation.java]: same here. "response" doesn't seem like the right term. More like "callback"?

File: ambry-router/src/main/java/com.github.ambry.router/GetBlobInfoOperation.java [hunk 3]
```diff
@@ -335,27 +337,32 @@ private void handleBody(InputStream payload, MessageMetadata messageMetadata)
...
+              progressTracker.setDecryptionSuccess();
+            } else {
+              decryptJobMetricsTracker.incrementOperationError();
+              logger.trace("Exception {} thrown on decryption for {}", exception, blobId);
+              setOperationException(
+                  new RouterException("Exception thrown on decrypting the content for " + blobId, exception,
+                      RouterErrorCode.UnexpectedInternalError));
+              progressTracker.setDecryptionFailed();
+            }
+            decryptJobMetricsTracker.completeResponseProcessing();
```
Comment (vgkholla) [inline ambry-router/src/main/java/com.github.ambry.router/GetBlobInfoOperation.java]: shouldn't this be after the `onPollReady()`? That's when the processing is truly complete

File: ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java [hunk 4]
```diff
@@ -455,6 +455,8 @@ private void updateChunkingAndSizeMetricsOnSuccessfulGet() {
     protected final Map<Integer, GetRequestInfo> correlationIdToGetRequestInfo = new TreeMap<>();
     // the state of the chunk.
     protected volatile ChunkState state;
+    // metrics tracker to track decrypt jobs
+    protected volatile CryptoJobMetricsTracker decryptJobMetricsTracker;
```
Comment (vgkholla) [inline ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java]: why `volatile`?

File: ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java [hunk 5]
```diff
@@ -1145,8 +1157,10 @@ private void handleSimpleBlob(BlobData blobData, byte[] userMetadata, ByteBuffer
           long startTimeMs = System.currentTimeMillis();
           decryptCallbackResultInfo = new DecryptCallBackResultInfo();
           progressTracker.initializeDecryptionTracker();
+          decryptJobMetricsTracker = new CryptoJobMetricsTracker(routerMetrics.decryptJobMetrics);
```
Comment (vgkholla) [inline ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java]: need to understand why these are created at multiple places - a little hard to follow

File: ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouterMetrics.java [hunk 6]
```diff
@@ -622,3 +630,126 @@ void trackAgeAtAccess(long creationTimeMs) {
...
+    roundTripTimeInMs = metricRegistry.histogram(MetricRegistry.name(ownerClass, requestType + ROUND_TRIP_TIME_SUFFIX));
+
+    operationRate = metricRegistry.meter(MetricRegistry.name(ownerClass, requestType + OPERATION_RATE_SUFFIX));
+    operationErrorRate = metricRegistry.meter(
+        MetricRegistry.name(ownerClass, requestType + OPERATION_ERROR_SUFFIX + OPERATION_RATE_SUFFIX));
+  }
+}
+
+/**
+ * Construct to support end-to-end metrics tracking for crypto jobs. Usually accompanies a single
```
Comment (vgkholla) [inline ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouterMetrics.java]: the doc is a little confusing - what do you mean by one-to-one mapping? Multiple instances of these trackers seem to be created in the operations?

File: ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouterMetrics.java [hunk 7]
```diff
@@ -622,3 +630,126 @@ void trackAgeAtAccess(long creationTimeMs) {
...
+  CryptoJobMetricsTracker(CryptoJobMetrics cryptoJobMetrics) {
+    this.cryptoJobMetrics = cryptoJobMetrics;
+  }
+
+  /**
+   * Starts the operation tracker for the corresponding CryptoJob
+   */
+  void startTracker() {
+    trackerStartTimeMs = System.currentTimeMillis();
+    lastSeenTimeMs = trackerStartTimeMs;
```
Comment (vgkholla) [inline ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouterMetrics.java]: why is this set here?

File: ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouterMetrics.java [hunk 8]
```diff
@@ -622,3 +630,126 @@ void trackAgeAtAccess(long creationTimeMs) {
...
+    lastSeenTimeMs = curTimeMs;
+  }
+
+  /**
+   * Notifies that the response processing has started for the corresponding CryptoJob
+   */
+  void startResponseProcessing() {
+    long curTimeMs = System.currentTimeMillis();
+    cryptoJobMetrics.responseProcessingWaitTimeInMs.update(curTimeMs - lastSeenTimeMs);
+    lastSeenTimeMs = curTimeMs;
```
Comment (vgkholla) [inline ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouterMetrics.java]: IMO, using this "lastSeenTime" is very confusing. Why not have "requestProcessingStartTime", "responseProcessingStartTime" etc? (also I don't think "request" and "response" is the correct terminology)

File: ambry-router/src/main/java/com.github.ambry.router/GetBlobInfoOperation.java [hunk 9]
```diff
@@ -54,6 +54,8 @@
   private final ProgressTracker progressTracker;
   // refers to blob properties received from the server
   private BlobProperties serverBlobProperties;
+  // metrics tracker to track decrypt jobs
+  private volatile CryptoJobMetricsTracker decryptJobMetricsTracker;
```
[the remark this answers is not in the document]
Comment (nsivabalan) [reply ambry-router/src/main/java/com.github.ambry.router/GetBlobInfoOperation.java]: Its been used by two diff threads (request response handler and crypto job handler) and hence.

File: ambry-router/src/main/java/com.github.ambry.router/GetBlobInfoOperation.java (hunk 3 above)
Comment (nsivabalan) [reply ambry-router/src/main/java/com.github.ambry.router/GetBlobInfoOperation.java]: I felt it doesn't make much of a diff. Will change it.

File: ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java (hunk 5 above)
Comment (nsivabalan) [reply ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java]: New CryptoJobMetricsTracker will be created for every encrypt or decrypt job and hence. For Puts, encryption happens only at one place (one method), where as decryption happens at 3 places (data chunk, simple and metadata chunk) and hence.

File: ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouterMetrics.java (hunk 6 above)
Comment (nsivabalan) [reply ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouterMetrics.java]: yes. check out my previous comment.

File: ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouterMetrics.java (hunk 7 above)
Comment (nsivabalan) [reply ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouterMetrics.java]: what do you mean by set here. Sorry I don't get you. Do you suggest it to be set elsewhere?

**Comment by nsivabalan** [comment]
@vgkholla : have addressed your comments.

**Comment by nsivabalan** [comment]
@vgkholla : reminder to review

File: ambry-router/src/main/java/com.github.ambry.router/PutOperation.java [hunk 10]
```diff
@@ -646,6 +646,10 @@ void setOperationExceptionAndComplete(Exception exception) {
     private int failedAttempts;
     // the partitionId chosen for the current chunk.
     private PartitionId partitionId;
+    // metrics tracker to track encrypt jobs
+    private CryptoJobMetricsTracker encryptJobMetricsTracker =
+        new CryptoJobMetricsTracker(routerMetrics.encryptJobMetrics);
+    ;
```
Comment (vgkholla) [inline ambry-router/src/main/java/com.github.ambry.router/PutOperation.java]: minor: extra semicolon

File: ambry-router/src/main/java/com.github.ambry.router/PutOperation.java [hunk 11]
```diff
@@ -646,6 +646,10 @@ void setOperationExceptionAndComplete(Exception exception) {
     private int failedAttempts;
     // the partitionId chosen for the current chunk.
     private PartitionId partitionId;
+    // metrics tracker to track encrypt jobs
+    private CryptoJobMetricsTracker encryptJobMetricsTracker =
```
Comment (vgkholla) [inline ambry-router/src/main/java/com.github.ambry.router/PutOperation.java]: `final`?

File: ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouterMetrics.java [hunk 12]
```diff
@@ -622,3 +630,150 @@ void trackAgeAtAccess(long creationTimeMs) {
...
+   */
+  void onJobSubmission() {
+    jobSubmissionTimeMs = System.currentTimeMillis();
+  }
+
+  /**
+   * Notifies that the job processing has started for the corresponding CryptoJob
+   */
+  void onJobProcessingStart() {
+    cryptoJobMetrics.operationRate.mark();
```
[the remark this answers is not in the document]
Comment (nsivabalan) [reply ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouterMetrics.java]: makes sense. will fix it

**Comment by nsivabalan** [comment]
@vgkholla : have addressed the last few comments
